candle-binding/src/unified_classifier.rs

```diff
@@ -0,0 +1,431 @@
+// Unified Classifier for Batch Inference Support
+// This module implements a unified classification system that:
+// 1. Uses a single shared ModernBERT encoder for all tasks
+// 2. Supports true batch inference (multiple texts in one forward pass)
+// 3. Provides multiple task heads (intent, PII, security) with shared backbone
+// 4. Eliminates memory waste from multiple model instances
+
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+use anyhow::{Error as E, Result};
+use candle_core::{Device, IndexOp, Tensor};
+use candle_nn::{Linear, Module};
+use candle_transformers::models::modernbert::{Config, ModernBert};
+use serde_json;
+use tokenizers::{Encoding, PaddingParams, PaddingStrategy, Tokenizer};
+
+/// Unified classification result for a single text
+#[derive(Debug, Clone)]
+pub struct UnifiedClassificationResult {
+    pub intent_result: IntentResult,
+    pub pii_result: PIIResult,
+    pub security_result: SecurityResult,
+}
+
+/// Intent classification result
+#[derive(Debug, Clone)]
+pub struct IntentResult {
+    pub category: String,
+    pub confidence: f32,
+    pub probabilities: Vec<f32>,
+}
+
+/// PII detection result
+#[derive(Debug, Clone)]
+pub struct PIIResult {
+    pub has_pii: bool,
+    pub pii_types: Vec<String>,
+    pub confidence: f32,
+}
+
+/// Security detection result
+#[derive(Debug, Clone)]
+pub struct SecurityResult {
+    pub is_jailbreak: bool,
+    pub threat_type: String,
+    pub confidence: f32,
+}
+
+/// Batch classification results
+#[derive(Debug)]
+pub struct BatchClassificationResult {
+    pub intent_results: Vec<IntentResult>,
+    pub pii_results: Vec<PIIResult>,
+    pub security_results: Vec<SecurityResult>,
+    pub batch_size: usize,
+}
+
+/// Unified classifier with shared ModernBERT backbone and multiple task heads
+pub struct UnifiedClassifier {
+    // Shared ModernBERT encoder (saves ~800MB memory vs 3 separate models)
```

Comment:
@OneZero-Y we are thinking the same: using the same base bert model in all classification tasks. But in the current model training, the base bert are all different. So this is not working right now.

The future step is in the two directions:
- [Multi task fine tuning](https://github.com/vllm-project/semantic-router/tree/main/src/training/multitask_bert_fine_tuning): A single classification head for all classes. This attempt however, yields very poor accuracy.
- LoRA. We are still working on this. If you like to get started, that'll be very promising.

Reply:
That's great! This will need to do in both training and classification:
- The candle lora crate has some [bert examples](https://github.com/EricLBuehler/candle-lora/blob/master/candle-lora-transformers/examples/bert.rs) but no modernbert. If you can get this started, that'll be cool!
- The [fine tuning scripts](https://github.com/vllm-project/semantic-router/tree/main/src/training) will use the PEFT library for the lora support.
- Then the candle-binding can be ready to support multi lora classification tasks.

Reply:
@OneZero-Y great work!